### PR TITLE
[CLIPPER-115][CLIPPER-116] Get models and containers

### PR DESCRIPTION
**Pull request by dcrankshaw**

### CODE REVIEW

File: management/clipper_manager.py [hunk 1]
```diff
@@ -334,7 +334,119 @@ def get_app_info(self, name):
...
+        ----------
+        verbose : bool
+            If set to False, the returned list contains the apps' names.
+            If set to True, the list contains model info dictionaries.
+
+        Returns
+        -------
+        list
+            Returns a list of information about all apps registered to Clipper.
+            If no modelss are registered with Clipper, an empty list is returned.
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: s\modelss\models

File: management/clipper_manager.py [hunk 2]
```diff
@@ -334,7 +334,119 @@ def get_app_info(self, name):
...
+            The name of the model to look up
+
+        Returns
+        -------
+        dict
+            Returns a dictionary with the specified model's info.
+            If no model with name `name` is registered with Clipper, None is returned.
+        """
+        url = "http://%s:1338/admin/get_model" % self.host
+        req_json = json.dumps({"name": name})
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Request json is incorrect. This should conform to the schema defined in the management frontend:

```
   {
    "model_name" := string,
    "model_version" := int,
   }
```

This method should also have a `version` parameter with documentation.

File: management/clipper_manager.py [hunk 3]
```diff
@@ -334,7 +334,119 @@ def get_app_info(self, name):
...
+            return None
+
+    def get_container_info(self, model_name, model_version, replica_id):
+        """Gets detailed information about a registered container.
+
+        Parameters
+        ----------
+        name : str
+            The name of the container to look up
+
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Add parameter info for `model_version`, `replica_id`

File: management/clipper_manager.py [hunk 4]
```diff
@@ -334,7 +334,119 @@ def get_app_info(self, name):
...
+        else:
+            print(r.text)
+            return None
+
+    def get_container_info(self, model_name, model_version, replica_id):
+        """Gets detailed information about a registered container.
+
+        Parameters
+        ----------
+        name : str
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: s\name\model_name

File: management/clipper_manager.py [hunk 5]
```diff
@@ -334,7 +334,119 @@ def get_app_info(self, name):
...
+            The name of the container to look up
+
+        Returns
+        -------
+        dict
+            Returns a dictionary with the specified container's info.
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Nit: Remove `Returns` from description

File: src/management/src/management_frontend.hpp [hunk 6]
```diff
@@ -244,6 +276,90 @@ class RequestHandler {
...
+        [this](std::shared_ptr<HttpServer::Response> response,
+               std::shared_ptr<HttpServer::Request> request) {
+          try {
+            clipper::log_info(LOGGING_TAG_MANAGEMENT_FRONTEND,
+                              "Get model info POST request");
+            std::string result = get_container(request->content.string());
+            respond_http(result, "200 OK", response);
+          } catch (const json_parse_error& e) {
+            std::string err_msg =
+                json_error_msg(e.what(), GET_MODEL_REQUESTS_SCHEMA);
```
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: This should be `GET_CONTAINER_REQUESTS_SCHEMA`

File: src/management/src/management_frontend.hpp [hunk 7]
```diff
@@ -244,6 +276,90 @@ class RequestHandler {
...
+            respond_http(err_msg, "400 Bad Request", response);
+          } catch (const json_semantic_error& e) {
+            std::string err_msg =
+                json_error_msg(e.what(), GET_MODEL_REQUESTS_SCHEMA);
```
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: `GET_CONTAINER_REQUESTS_SCHEMA`

File: src/libclipper/include/clipper/redis.hpp [hunk 8]
```diff
@@ -169,6 +156,14 @@ std::unordered_map<std::string, std::string> get_model_by_key(
 std::vector<std::string> get_all_model_names(redox::Redox& redis);
 
 /**
+ * Looks up models listed in the model table.
+ *
+ * \return Returns a vector of models and versions. If no models
+ * were found, then an empty vector will be returned.
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/redis.hpp]: Nit: s\were\are

File: src/libclipper/test/json_util_test.cpp [hunk 9]
```diff
@@ -249,18 +249,64 @@ TEST_F(SetJsonDocTest, TestSetJsonDocFromRedisAppMetadata) {
...
+                        container_name, model_path));
+
+  std::unordered_map<std::string, std::string> model_metadata =
+      get_model(*redis_, model);
+
+  rapidjson::Document d;
+  redis_model_metadata_to_json(d, model_metadata);
+
+  ASSERT_EQ(get_string(d, "input_type"), input_type);
+
```
Comment (Corey-Zumar) [inline src/libclipper/test/json_util_test.cpp]: Nit: unnecessary blank line

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
Found some naming issues in clipper manager and the management frontend that should be easy to correct. Other than that and a couple formatting nits, it all looks good!

File: management/clipper_manager.py (hunk 2 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Uggh. Good catch. Fixed.

File: management/clipper_manager.py (hunk 4 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

File: management/clipper_manager.py (hunk 3 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

File: management/clipper_manager.py (hunk 5 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

File: src/management/src/management_frontend.hpp (hunk 6 above)
Comment (dcrankshaw) [reply src/management/src/management_frontend.hpp]: Done

File: src/management/src/management_frontend.hpp (hunk 7 above)
Comment (dcrankshaw) [reply src/management/src/management_frontend.hpp]: Done

File: src/libclipper/include/clipper/redis.hpp (hunk 8 above)
Comment (dcrankshaw) [reply src/libclipper/include/clipper/redis.hpp]: Done

File: src/libclipper/test/json_util_test.cpp (hunk 9 above)
Comment (dcrankshaw) [reply src/libclipper/test/json_util_test.cpp]: Removed

File: src/management/src/management_frontend.hpp (hunk 7 above)
Comment (dcrankshaw) [reply src/management/src/management_frontend.hpp]: Oh. This is in GET_MODEL so this one is okay.

**Comment by dcrankshaw** [comment]
@Corey-Zumar Addressed your review comments

File: management/clipper_manager.py [hunk 10]
```diff
@@ -371,22 +372,28 @@ def get_all_models(self, verbose=False):
...
         dict
             Returns a dictionary with the specified model's info.
-            If no model with name `name` is registered with Clipper, None is returned.
+            If no model with name `model_name@model_version` is
+            registered with Clipper, None is returned.
         """
         url = "http://%s:1338/admin/get_model" % self.host
-        req_json = json.dumps({"name": name})
+        req_json = json.dumps({
+            "model_name": name,
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: This should be `"model_name": model_name,`

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
One more naming error. Can approve and merge after

**Comment by dcrankshaw** [comment]
@Corey-Zumar fixed naming error

**Review by Corey-Zumar — APPROVED** [review]
LGTM

**Pull request merged**